libgit/repo.go

```diff
@@ -545,3 +571,167 @@ func DeleteRepo(
+	}
+	_, err = f.Seek(0, io.SeekStart)
+	if err != nil {
+		return err
+	}
+	err = f.Truncate(0)
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(buf)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RenameRepo renames the repo from an old name to a new name.  It
+// leaves a symlink behind so that old remotes will continue to work.
+// The caller is responsible for syncing the FS and flushing the
+// journal, if desired.
+func RenameRepo(
+	ctx context.Context, config libkbfs.Config, tlfHandle *libkbfs.TlfHandle,
+	oldRepoName, newRepoName string) error {
+	kbfsOps := config.KBFSOps()
+	rootNode, _, err := kbfsOps.GetOrCreateRootNode(
+		ctx, tlfHandle, libkbfs.MasterBranch)
+	if err != nil {
+		return err
+	}
+	normalizedOldRepoName := normalizeRepoName(oldRepoName)
+	normalizedNewRepoName := normalizeRepoName(newRepoName)
+
+	repoNode, _, err := kbfsOps.Lookup(ctx, rootNode, kbfsRepoDir)
+	if err != nil {
+		return err
+	}
+
+	// Does the old repo definitely exist?
+	_, _, err = kbfsOps.Lookup(ctx, repoNode, normalizedOldRepoName)
+	if err != nil {
+		return err
+	}
+
+	if oldRepoName == newRepoName {
+		// The names are the same, nothing else to do.
+		return nil
+	}
+
+	fs, err := libfs.NewFS(ctx, config, tlfHandle, path.Join(kbfsRepoDir), "")
+	if err != nil {
+		return err
+	}
+
+	oldRepoFS, err := fs.Chroot(normalizedOldRepoName)
+	if err != nil {
+		return err
+	}
+
+	// Take locks in both repos during rename (same lock that's taken
+	// for new repo creation).
+	oldLockFile, err := takeConfigLock(
+		oldRepoFS.(*libfs.FS), tlfHandle, oldRepoName)
+	if err != nil {
+		return err
+	}
+	defer oldLockFile.Close()
+
+	if normalizedOldRepoName == normalizedNewRepoName {
+		// All we need to do is update the name in the config file,
+		// and the MD.
+		err = renameRepoInConfigFile(ctx, oldRepoFS, newRepoName)
+		if err != nil {
+			return err
+		}
+		return UpdateRepoMD(ctx, config, tlfHandle, oldRepoFS)
+	}
+
+	// Does the new repo not exist yet?
+	_, _, err = kbfsOps.Lookup(ctx, repoNode, normalizedNewRepoName)
+	switch errors.Cause(err).(type) {
+	case libkbfs.NoSuchNameError:
+		// The happy path.
+	case nil:
+		newRepoFS, err := fs.Chroot(normalizedNewRepoName)
+		if err != nil {
+			return err
+		}
+		return makeExistingRepoError(ctx, config, newRepoFS, newRepoName)
+	default:
+		return err
+	}
+
+	// Make the new repo subdir just so we can take the lock inside
+	// the new repo.  (We'll delete the new dir after the rename.)
+	err = fs.MkdirAll(normalizedNewRepoName, 0777)
+	if err != nil {
+		return err
+	}
+	newRepoFS, err := fs.Chroot(normalizedNewRepoName)
+	if err != nil {
+		return err
+	}
+	newLockFile, err := takeConfigLock(
+		newRepoFS.(*libfs.FS), tlfHandle, newRepoName)
+	if err != nil {
+		return err
+	}
+	defer newLockFile.Close()
+
+	// Rename this new dir out of the way before we rename.
+	fi, err := fs.Stat(normalizedNewRepoName)
+	if err != nil {
+		return err
+	}
+	err = recursiveDelete(ctx, fs, fi)
```

Comment:
This would delete what's referenced by `newLockFile` too right? And since it's open already, we can still call `Close()` to unlock it?

Reply:
Yes, it would delete the locked file too, but that's fine because we still have a file handle open to it.  And yes Close should still unlock it.  I don't see another way to lock down the repo's new name before doing the rename.

Reply:
OK sounds good. I was just making sure I understood correctly. Thanks!